src/services/CapabilitiesManager.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import { getCapabilities as _getCapabilities } from '@nextcloud/capabilities'
+import { showError, TOAST_PERMANENT_TIMEOUT } from '@nextcloud/dialogs'
+
+import { getRemoteCapabilities } from './federationService.ts'
+import BrowserStorage from '../services/BrowserStorage.js'
+import type { Capabilities, JoinRoomFullResponse } from '../types'
+
+interface ICapabilitiesManager {
+	localCapabilities: Capabilities
+	remoteCapabilities: Record<string, Capabilities & Partial<{ hash: string }>>
+	getCapabilities: (token: 'local'|string) => Capabilities
+}
+
+type Config = Capabilities['spreed']['config']
+
+class CapabilitiesManager implements ICapabilitiesManager {
+
+	public localCapabilities
+
+	public remoteCapabilities
+
+	static instance: ICapabilitiesManager
+
+	constructor() {
+		this.localCapabilities = _getCapabilities() as Capabilities
+		this.remoteCapabilities = this.restoreRemoteCapabilities()
+	}
+
+	public getCapabilities(token: 'local'|string = 'local') {
+		return (token === 'local' || !this.remoteCapabilities[token])
+			? this.localCapabilities
+			: this.remoteCapabilities[token]
+	}
+
+	public hasTalkFeature(token: 'local'|string, feature: string): boolean {
+		if (this.localCapabilities.spreed['features-local'].includes(feature)) {
+			return this.localCapabilities?.spreed?.features?.includes(feature) ?? false
+		}
+		return this.getCapabilities(token)?.spreed?.features?.includes(feature) ?? false
+	}
```

Comment:
If the feature is local, we don't need to check remote server at all because those features are usually not affecting the chat, call etc (e.g: note-to-self)
If the feature is not flagged as local, both remote and local capabilities should be checked using `AND` (if any of the server doesn't have that feature, it should not be there).

Note: we need to be careful when flagging feature/config with local, I noticed that `config => attachments => allowed` is local whereas it is affecting the chat.

Reply:
I'd agree that feature (and also config) should be respected from both local / remote, if that's related to the conversation; remote user shouldn't have more possibilities, than he already has locally, and more than a local user:

Remote cap. | Local cap. | Resulting | Example
-- | -- | -- | --
🔴 | 🔴 | 🔴 | Not conflicting
🟢 | 🟢 | 🟢 | Not conflicting
🟢 | 🔴 | 🔴 | "Moderator" from 29 can't ban participants, because he isn't aware of the feature 
🔴 | 🟢 | 🔴 | Can't upload attachments, if not allowed on remote server

Reply:
> But for the config, i think remote should be respected, unless we define it in config-local: as we have numeric and string values there, not only boolean

For config, it's a bit tricky, maybe min(remote, local) for numeric config and for not measurable numeric (like integers to reflect modes), use remote, right  ?
For now, `chat-read-status` (local) which should be following remote server, no ?

Reply:
Or we treat each config individually for now.
chat-read-status differs per user and set up by him, if I'm not mistaken, so should be from local

Reply:
> * `chat-read-status` (local) - On conversation API v3 and the chat API the last common read message is exposed which can be used to update the "read status" flag of own chat messages. The info should be shown only when the user also shares their read status. The user's value can be found in `config => chat => read-privacy`.

That means it is first configured by server (expose  the last common read message) and then enabled/disabled by user.

Reply:
`chat-read-status` isn't used in frontend code, so consider this out of scope of this PR. Let's stick with:
- features should be respected on both
- configs should be respected from remote, unless we don't define them as local || unless we don't set code to respect both individually for each case

Speaking of which, @nickvergessen :
`config => attachments => allowed`, shouldn't we respect both local and remote?
`features => avatar`, aren't we proxying the same local endpoint for avatar, so only local is enough?

Reply:
yeah the docs might be from before we decided to proxy avatars.

As for attachments, I guess we should check both, but we can clarify once we implement them